app/ydoc-shared/src/ast/parse.ts

```diff
@@ -512,166 +549,117 @@ export function print(ast: Ast): PrintedSource {
     nodes: new Map(),
     tokens: new Map(),
   }
-  const code = ast.printSubtree(info, 0, undefined)
+  const code = ast.printSubtree(info, 0, null)
   return { info, code }
 }
 
 /**
- * Used by `Ast.printSubtree`. Note that some AST types have overrides.
+ * Used by `Ast.printSubtree`.
  * @internal
  */
 export function printAst(
   ast: Ast,
   info: SpanMap,
   offset: number,
-  parentIndent: string | undefined,
-  verbatim?: boolean,
+  parentIndent: string | null,
+  verbatim: boolean = false,
 ): string {
   let code = ''
-  for (const child of ast.concreteChildren(verbatim)) {
+  let currentLineIndent = parentIndent
+  let prevIsNewline = false
+  let isFirstToken = offset === 0
+  for (const child of ast.concreteChildren({ verbatim, indent: parentIndent })) {
     if (!isTokenId(child.node) && ast.module.get(child.node) === undefined) continue
-    if (child.whitespace != null) {
-      code += child.whitespace
-    } else if (code.length != 0) {
-      code += ' '
+    if (prevIsNewline) currentLineIndent = child.whitespace
+    const token = isTokenId(child.node) ? ast.module.getToken(child.node) : undefined
+    if (
+      token?.tokenType_ == RawAst.Token.Type.Newline &&
+      isFirstToken &&
+      (!token.code_ || token.code_ === '\n')
+    ) {
```

Comment:
Why do we check both token type and code? Are there newline tokens which are not `\n`?

Reply:
Added a comment.